File: 5_WebApps/KhachLink/Services/EnhancedCartService.cs
using VanAn.Shared.Domain;
using VanAn.Shared.Services;
using VanAn.KhachLink.Models;

namespace VanAn.KhachLink.Services
{
    public interface IEnhancedCartService
    {
        Task<bool> AddItemAsync(CartItem item);
        Task<bool> RemoveItemAsync(string productId);
        Task<bool> UpdateQuantityAsync(string productId, int quantity);
        Task<List<CartItem>> GetItemsAsync();
        Task<decimal> GetTotalAsync();
        Task<bool> ClearCartAsync();
        Task<SyncResult> SyncCartAsync();
        Task<bool> SaveCartOfflineAsync();
        Task<bool> LoadCartOfflineAsync();
    }

    public class EnhancedCartService(
        IIndexedDBService indexedDBService,
        ICartService cartService,
        ISyncConflictResolver conflictResolver,
        ILogger<EnhancedCartService> logger) : IEnhancedCartService
    {
        private readonly IIndexedDBService _indexedDBService = indexedDBService;
        private readonly ICartService _cartService = cartService;
        private readonly ISyncConflictResolver _conflictResolver = conflictResolver;
        private readonly ILogger<EnhancedCartService> _logger = logger;
        private readonly string _cartKey = "user_cart";

        public async Task<bool> AddItemAsync(CartItem item)
        {
            try
            {
                // Add to local cart
                List<CartItem> items = await GetItemsAsync();
                CartItem? existingItem = items.FirstOrDefault(i => i.Id == item.Id);

                if (existingItem != null)
                {
                    // CartItem is immutable, create new instance with updated quantity
                    int index = items.IndexOf(existingItem);
                    items[index] = existingItem with { Quantity = existingItem.Quantity + item.Quantity };
                }
                else
                {
                    // Add new item
                    items.Add(item);
                }

                // Save to IndexedDB
                await _indexedDBService.SetAsync(_cartKey, items);

                // Try to sync immediately if online
                _ = Task.Run(SyncCartAsync);

                _logger.LogInformation("Item added to cart: {ProductId}, Quantity: {Quantity}", item.Id, item.Quantity);
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to add item to cart: {ProductId}", item.Id);
                return false;
            }
        }

        public async Task<bool> RemoveItemAsync(string productId)
        {
            try
            {
                List<CartItem> items = await GetItemsAsync();
                CartItem? item = items.FirstOrDefault(i => i.Id.ToString() == productId);

                if (item != null)
                {
                    items.Remove(item);
                    await _indexedDBService.SetAsync(_cartKey, items);

                    // Try to sync immediately if online
                    _ = Task.Run(SyncCartAsync);

                    _logger.LogInformation("Item removed from cart: {ProductId}", productId);
                    return true;
                }

                return false;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to remove item from cart: {ProductId}", productId);
                return false;
            }
        }

        public async Task<bool> UpdateQuantityAsync(string productId, int quantity)
        {
            try
            {
                if (quantity <= 0)
                {
                    return await RemoveItemAsync(productId);
                }

                List<CartItem> items = await GetItemsAsync();
                CartItem? item = items.FirstOrDefault(i => i.Id.ToString() == productId);

                if (item != null)
                {
                    // CartItem is immutable, create new instance with updated quantity
                    int index = items.IndexOf(item);
                    items[index] = item with { Quantity = quantity };

                    await _indexedDBService.SetAsync(_cartKey, items);

                    // Try to sync immediately if online
                    _ = Task.Run(SyncCartAsync);

                    _logger.LogInformation("Cart item quantity updated: {ProductId}, New Quantity: {Quantity}", productId, quantity);
                    return true;
                }

                return false;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to update cart item quantity: {ProductId}", productId);
                return false;
            }
        }

        public async Task<List<CartItem>> GetItemsAsync()
        {
            try
            {
                List<CartItem>? items = await _indexedDBService.GetAsync<List<CartItem>>(_cartKey);
                return items ?? [];
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to get cart items");
                return [];
            }
        }

        public async Task<decimal> GetTotalAsync()
        {
            try
            {
                List<CartItem> items = await GetItemsAsync();
                return items.Sum(i => i.TotalPrice);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to calculate cart total");
                return 0;
            }
        }

        public async Task<bool> ClearCartAsync()
        {
            try
            {
                await _indexedDBService.RemoveAsync(_cartKey);

                // Try to sync immediately if online
                _ = Task.Run(_cartService.ClearCartAsync);

                _logger.LogInformation("Cart cleared");
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to clear cart");
                return false;
            }
        }

        public async Task<SyncResult> SyncCartAsync()
        {
            SyncResult result = new();

            try
            {
                // Get offline cart
                List<CartItem> offlineItems = await GetItemsAsync();
                if (offlineItems.Count == 0)
                {
                    result.Success = true;
                    result.ErrorMessage = "No items to sync";
                    return result;
                }

                // Get server cart
                List<CartItem> serverItems = await _cartService.GetItemsAsync();

                // Resolve conflicts
                ConflictResolution conflictResolution = await _conflictResolver.ResolveCartConflictAsync(
                    offlineItems.Select(ToOfflineItemDto).ToList(),
                    serverItems);

                if (!conflictResolution.Success)
                {
                    result.Success = false;
                    result.ErrorMessage = conflictResolution.Reason;
                    return result;
                }

                // Sync resolved items to server
                if (conflictResolution.Action == ResolutionAction.Merge && conflictResolution.MergedOrder != null)
                {
                    // Clear server cart first
                    await _cartService.ClearCartAsync();

                    // Add merged items to server
                    foreach (OfflineOrderItemDto item in conflictResolution.MergedOrder.Items)
                    {
                        await _cartService.AddItemAsync(new CartItem
                        {
                            Id = Guid.NewGuid(),
                            ProductId = Guid.Parse(item.ProductId),
                            ProductName = item.ProductName ?? string.Empty,
                            Description = string.Empty,
                            Quantity = item.Quantity,
                            UnitPrice = item.UnitPrice
                        });
                    }

                    // Update offline cart with merged items
                    List<CartItem> mergedCartItems = conflictResolution.MergedOrder.Items.Select(ToCartItem).ToList();
                    await _indexedDBService.SetAsync(_cartKey, mergedCartItems);
                }

                result.Success = true;
                _logger.LogInformation("Cart synced successfully with {ItemCount} items", offlineItems.Count);
            }
            catch (Exception ex)
            {
                result.Success = false;
                result.ErrorMessage = ex.Message;
                _logger.LogError(ex, "Failed to sync cart");
            }

            return result;
        }

        public async Task<bool> SaveCartOfflineAsync()
        {
            try
            {
                List<CartItem> items = await _cartService.GetItemsAsync();
                await _indexedDBService.SetAsync(_cartKey, items);

                _logger.LogInformation("Cart saved offline with {ItemCount} items", items.Count);
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to save cart offline");
                return false;
            }
        }

        public async Task<bool> LoadCartOfflineAsync()
        {
            try
            {
                List<CartItem>? offlineItems = await _indexedDBService.GetAsync<List<CartItem>>(_cartKey);
                if (offlineItems == null || offlineItems.Count == 0)
                {
                    return true; // No offline cart to load
                }

                // Clear current server cart
                await _cartService.ClearCartAsync();

                // Load offline items to server cart
                foreach (CartItem item in offlineItems)
                {
                    await _cartService.AddItemAsync(item);
                }

                _logger.LogInformation("Cart loaded from offline with {ItemCount} items", offlineItems.Count);
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to load cart from offline");
                return false;
            }
        }

        private OfflineOrderItemDto ToOfflineItemDto(CartItem item)
        {
            return new OfflineOrderItemDto
            {
                ProductId = item.ProductId.ToString(),
                Quantity = item.Quantity,
                UnitPrice = item.UnitPrice,
                TotalPrice = item.TotalPrice,
                ProductName = item.ProductName
            };
        }

        private CartItem ToCartItem(OfflineOrderItemDto item)
        {
            return new CartItem
            {
                Id = Guid.NewGuid(),
                ProductId = Guid.Parse(item.ProductId),
                ProductName = item.ProductName ?? string.Empty,
                Description = string.Empty,
                Quantity = item.Quantity,
                UnitPrice = item.UnitPrice
            };
        }
    }
}

File: 5_WebApps/KhachLink/Services/EnhancedCartService.ts
import { IndexedDBService, IndexedDBServiceImpl } from './IndexedDBService';

export interface CartService {
    addItem(item: CartItem): Promise<void>;
    removeItem(productId: string): Promise<void>;
    updateQuantity(productId: string, quantity: number): Promise<void>;
    getItems(): Promise<CartItem[]>;
    getTotal(): Promise<number>;
    clearCart(): Promise<void>;
    syncCart(): Promise<SyncResult>;
    saveCartOffline(): Promise<void>;
    loadCartOffline(): Promise<void>;
}

export class EnhancedCartServiceImpl implements CartService {
    private indexedDB: IndexedDBService;
    private isOnline: boolean = navigator.onLine;
    private cartKey = 'user_cart';
    
    constructor() {
        this.indexedDB = new IndexedDBServiceImpl();
        
        // Monitor online/offline status
        window.addEventListener('online', () => {
            this.isOnline = true;
            this.syncCart();
        });
        
        window.addEventListener('offline', () => {
            this.isOnline = false;
        });
    }
    
    async addItem(item: CartItem): Promise<void> {
        try {
            // Get current cart
            const items = await this.getItems();
            const existingItem = items.find(i => i.productId === item.productId);
            
            if (existingItem) {
                // Update quantity
                existingItem.quantity += item.quantity;
                existingItem.totalPrice = existingItem.quantity * existingItem.unitPrice;
            } else {
                // Add new item
                items.push(item);
            }
            
            // Save to IndexedDB
            await this.indexedDB.init();
            await this.indexedDB.storeCart({
                id: this.cartKey,
                items: items,
                updatedAt: new Date().toISOString()
            });
            
            // Try to sync immediately if online
            if (this.isOnline) {
                await this.syncCart();
            }
        } catch (error) {
            console.error('Error adding item to cart:', error);
            throw error;
        }
    }
    
    async removeItem(productId: string): Promise<void> {
        try {
            const items = await this.getItems();
            const item = items.find(i => i.productId === productId);
            
            if (item) {
                items.splice(items.indexOf(item), 1);
                
                await this.indexedDB.init();
                await this.indexedDB.storeCart({
                    id: this.cartKey,
                    items: items,
                    updatedAt: new Date().toISOString()
                });
                
                // Try to sync immediately if online
                if (this.isOnline) {
                    await this.syncCart();
                }
            }
        } catch (error) {
            console.error('Error removing item from cart:', error);
            throw error;
        }
    }
    
    async updateQuantity(productId: string, quantity: number): Promise<void> {
        try {
            if (quantity <= 0) {
                await this.removeItem(productId);
                return;
            }
            
            const items = await this.getItems();
            const item = items.find(i => i.productId === productId);
            
            if (item) {
                item.quantity = quantity;
                item.totalPrice = quantity * item.unitPrice;
                
                await this.indexedDB.init();
                await this.indexedDB.storeCart({
                    id: this.cartKey,
                    items: items,
                    updatedAt: new Date().toISOString()
                });
                
                // Try to sync immediately if online
                if (this.isOnline) {
                    await this.syncCart();
                }
            }
        } catch (error) {
            console.error('Error updating cart item quantity:', error);
            throw error;
        }
    }
    
    async getItems(): Promise<CartItem[]> {
        try {
            await this.indexedDB.init();
            const cart = await this.indexedDB.getCart(this.cartKey);
            return cart?.items || [];
        } catch (error) {
            console.error('Error getting cart items:', error);
            return [];
        }
    }
    
    async getTotal(): Promise<number> {
        try {
            const items = await this.getItems();
            return items.reduce((total, item) => total + item.totalPrice, 0);
        } catch (error) {
            console.error('Error calculating cart total:', error);
            return 0;
        }
    }
    
    async clearCart(): Promise<void> {
        try {
            await this.indexedDB.init();
            await this.indexedDB.storeCart({
                id: this.cartKey,
                items: [],
                updatedAt: new Date().toISOString()
            });
            
            // Try to sync immediately if online
            if (this.isOnline) {
                await this.syncCart();
            }
        } catch (error) {
            console.error('Error clearing cart:', error);
            throw error;
        }
    }
    
    async syncCart(): Promise<SyncResult> {
        const result: SyncResult = {
            success: false,
            syncedCount: 0,
            errorCount: 0,
            errors: []
        };
        
        if (!this.isOnline) {
            result.errors.push('Device is offline');
            return result;
        }
        
        try {
            await this.indexedDB.init();
            const cart = await this.indexedDB.getCart(this.cartKey);
            
            if (!cart || !cart.items.length) {
                result.success = true;
                result.errors.push('No items to sync');
                return result;
            }
            
            // Get server cart
            const serverItems = await this.getServerCart();
            
            // Resolve conflicts
            const mergedItems = this.mergeCartItems(cart.items, serverItems);
            
            // Sync merged items to server
            await this.syncToServer(mergedItems);
            
            // Update offline cart with merged items
            await this.indexedDB.storeCart({
                id: this.cartKey,
                items: mergedItems,
                updatedAt: new Date().toISOString()
            });
            
            result.success = true;
            result.syncedCount = mergedItems.length;
        } catch (error) {
            result.errorCount++;
            result.errors.push(`Cart sync failed: ${error}`);
        }
        
        return result;
    }
    
    async saveCartOffline(): Promise<void> {
        try {
            const serverItems = await this.getServerCart();
            
            await this.indexedDB.init();
            await this.indexedDB.storeCart({
                id: this.cartKey,
                items: serverItems,
                updatedAt: new Date().toISOString()
            });
        } catch (error) {
            console.error('Error saving cart offline:', error);
            throw error;
        }
    }
    
    async loadCartOffline(): Promise<void> {
        try {
            await this.indexedDB.init();
            const cart = await this.indexedDB.getCart(this.cartKey);
            
            if (!cart || !cart.items.length) {
                return; // No offline cart to load
            }
            
            // Clear server cart first
            await this.clearServerCart();
            
            // Load offline items to server cart
            for (const item of cart.items) {
                await this.addToServerCart(item);
            }
        } catch (error) {
            console.error('Error loading cart from offline:', error);
            throw error;
        }
    }
    
    private async getServerCart(): Promise<CartItem[]> {
        const response = await fetch('/api/cart');
        if (!response.ok) {
            throw new Error(`Failed to get server cart: ${response.statusText}`);
        }
        return await response.json();
    }
    
    private async syncToServer(items: CartItem[]): Promise<void> {
        // Clear server cart first
        await this.clearServerCart();
        
        // Add all items to server cart
        for (const item of items) {
            await this.addToServerCart(item);
        }
    }
    
    private async clearServerCart(): Promise<void> {
        const response = await fetch('/api/cart/clear', {
            method: 'POST'
        });
        if (!response.ok) {
            throw new Error(`Failed to clear server cart: ${response.statusText}`);
        }
    }
    
    private async addToServerCart(item: CartItem): Promise<void> {
        const response = await fetch('/api/cart/items', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(item)
        });
        if (!response.ok) {
            throw new Error(`Failed to add item to server cart: ${response.statusText}`);
        }
    }
    
    private mergeCartItems(offlineItems: CartItem[], serverItems: CartItem[]): CartItem[] {
        const mergedItems: CartItem[] = [];
        const processedProductIds = new Set<string>();
        
        // Add offline items first
        for (const offlineItem of offlineItems) {
            mergedItems.push(offlineItem);
            processedProductIds.add(offlineItem.productId);
        }
        
        // Add server items that aren't in offline
        for (const serverItem of serverItems) {
            if (!processedProductIds.has(serverItem.productId)) {
                mergedItems.push(serverItem);
            }
        }
        
        return mergedItems;
    }
}

export interface CartItem {
    productId: string;
    quantity: number;
    unitPrice: number;
    totalPrice: number;
}

export interface SyncResult {
    success: boolean;
    syncedCount: number;
    errorCount: number;
    errors: string[];
}

export interface CartStorage {
    id: string;
    items: CartItem[];
    updatedAt: string;
}
